metron-platform/metron-enrichment/src/main/java/org/apache/metron/threatintel/triage/ThreatTriageProcessor.java

```diff
@@ -52,15 +74,36 @@ public ThreatTriageProcessor( SensorEnrichmentConfig config
 
   @Nullable
   @Override
-  public Double apply(@Nullable Map input) {
-    List<Number> scores = new ArrayList<>();
+  public ThreatScore apply(@Nullable Map input) {
+
+    ThreatScore threatScore = new ThreatScore();
     StellarPredicateProcessor predicateProcessor = new StellarPredicateProcessor();
+    StellarProcessor processor = new StellarProcessor();
     VariableResolver resolver = new MapVariableResolver(input, sensorConfig.getConfiguration(), threatIntelConfig.getConfig());
+
+    // attempt to apply each rule to the threat
     for(RiskLevelRule rule : threatTriageConfig.getRiskLevelRules()) {
       if(predicateProcessor.parse(rule.getRule(), resolver, functionResolver, context)) {
-        scores.add(rule.getScore());
+
+        // add the rule's score to the overall threat score
+        String reason = execute(rule.getReason(), processor, resolver, String.class);
+        RuleScore score = new RuleScore(rule, reason);
+        threatScore.addRuleScore(score);
       }
     }
-    return threatTriageConfig.getAggregator().aggregate(scores, threatTriageConfig.getAggregationConfig());
+
+    // calculate the aggregate threat score
+    Aggregators aggregators = threatTriageConfig.getAggregator();
+    List<Number> allScores = threatScore.getRuleScores().stream().map(score -> score.getRule().getScore()).collect(Collectors.toList());
+    Double aggregateScore = aggregators.aggregate(allScores, threatTriageConfig.getAggregationConfig());
+
+    // return the overall threat score
+    threatScore.setScore(aggregateScore);
+    return threatScore;
+  }
+
+  private <T> T execute(String expression, StellarProcessor processor, VariableResolver resolver, Class<T> clazz) {
+    Object result = processor.parse(expression, resolver, functionResolver, context);
```

Comment:
This gets called to execute the 'reason' Stellar expression that is part of a threat triage rule.  It is only called on the rules that are 'fired' (rule predicate is true) for a particular message (line 89).  

A rule's 'reason' is something that could be changed by the user at any time.  I am unsure of a better way to do it.  Thoughts?